### PR TITLE
Couple of fixes to state migration

### PR DESCRIPTION
**Pull request by palango**
Fixes the following issues:

- #135 
- #160
- #157

### CODE REVIEW

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 1]
```diff
@@ -117,16 +94,16 @@ func applyStateMigrationChanges(genesis *core.Genesis, dbPath string, commit boo
 		Number:      migrationBlock,
 		GasLimit:    header.GasLimit,
 		GasUsed:     0,
-		Time:        uint64(time.Now().Unix()), // TODO(pl): Needed to avoid L1-L2 time mismatches
+		Time:        header.Time + 5,
 		Extra:       []byte("CeL2 migration"),
 		MixDigest:   common.Hash{},
 		Nonce:       types.BlockNonce{},
 		BaseFee:     new(big.Int).Set(header.BaseFee),
 	}
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go:103]: You need to add the empty withdrawals hash too!

**Comment by piersy** [comment]
Hey @palango why are we doing this?

* https://github.com/celo-org/optimism/issues/158

**Comment by palango** [comment]
> Hey @palango why are we doing this?
> 
> * [During migration, add existing balance when overwriting account #158](https://github.com/celo-org/optimism/issues/158)

Because otherwise we'd loose existing balance. We found out yesterday that some of the accounts from the OP namespace received payments in the past, so we need to handle this edge case.

File: op-chain-ops/cmd/celo-migrate/state.go (hunk 1 above)
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go:103]: Good point, added.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 2]
```diff
@@ -62,38 +65,6 @@ func applyStateMigrationChanges(genesis *core.Genesis, dbPath string, commit boo
 		return nil, fmt.Errorf("cannot open StateDB: %w", err)
 	}
 
-	// So far we applied changes in the memory VM and collected changes in the genesis struct
-	// Now we iterate through all accounts that have been written there and set them inside the statedb.
-	// This will change the state root
-	// Another property is that the total balance changes must be 0
```
Comment (alecps) [inline op-chain-ops/cmd/celo-migrate/state.go:68]: Why have we deleted this? I thought this was the meat of what the state migration script was supposed to be doing

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 3]
```diff
@@ -104,29 +75,36 @@ func applyStateMigrationChanges(genesis *core.Genesis, dbPath string, commit boo
 		return nil, err
 	}
 
-	// Create the header for the Bedrock transition block.
+	// Apply the changes to the state DB.
+	err = applyAllocsToState(db, genesis)
```
Comment (alecps) [inline op-chain-ops/cmd/celo-migrate/state.go]: Why have we moved this logic after the "Committing state DB" step?

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: Nice catch, that's unintentional.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 4]
```diff
@@ -208,3 +185,43 @@ func applyStateMigrationChanges(genesis *core.Genesis, dbPath string, commit boo
...
+				overwriteCounter++
+			}
+		}
+		db.CreateAccount(k)
+
+		db.SetNonce(k, v.Nonce)
+		db.SetBalance(k, balance)
+		db.SetCode(k, v.Code)
+		for key, value := range v.Storage {
+			db.SetState(k, key, value)
```
Comment (alecps) [inline op-chain-ops/cmd/celo-migrate/state.go]: Should we have counters / logging for times we overwrite storage or nonce?

**Review by alecps — APPROVED** [review]
Looks good!

**Comment by piersy** [comment]
@palango

Regarding this I'm still a bit confused

>Because otherwise we'd loose existing balance. We found out yesterday that some of the accounts from the OP namespace received payments in the past, so we need to handle this edge case.

So just to clarify, you are saying that some of the addresses of the contracts that we are inserting into our "L2 genesis" block have already received payments in the history of the celo L1?

If so I thought the previous approach was to fail?

**Comment by piersy** [comment]
>Regarding this I'm still a bit confused

>>Because otherwise we'd loose existing balance. We found out yesterday that some of the accounts from the OP namespace received payments in the past, so we need to handle this edge case.

>So just to clarify, you are saying that some of the addresses of the contracts that we are inserting into our "L2 genesis" block have already received payments in the history of the celo L1?

>If so I thought the previous approach was to fail?

Ok I have understood the thinking here. We need to maintain the correct amount of celo tokens. I was worried about what could be achieved by someone who had the private key for these accounts but as you mentioned we haven't seen any transactions issued from those accounts, even though they received funds and as Marek pointed out if someone had the key they likely would have used it on op or base.

File: op-chain-ops/cmd/celo-migrate/state.go (hunk 4 above)
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: I need to add a whitelist here as well, will move this out of this PR

**Pull request merged**